File: tests/docs/python/test_attribute_bindings.py
# Copyright (c) 2025-2026, NVIDIA CORPORATION. All rights reserved.
#
# NVIDIA CORPORATION and its licensors retain all intellectual property
# and proprietary rights in and to this software, related documentation
# and any modifications thereto.  Any use, reproduction, disclosure or
# distribution of this software and related documentation without an express
# license agreement from NVIDIA CORPORATION is strictly prohibited.

"""Tests for persistent attribute bindings, mapping, and CUDA write paths."""

from pathlib import Path

import numpy as np
import ovrtx
import ovstage
import pytest
import warp as wp

TEST_BASE_PATH = str((Path(__file__).parent / "../data/ovrtx-test-base.usda").resolve())


@wp.kernel
def _set_xform_translation_x(transforms: wp.array(dtype=wp.mat44d), x: wp.float64):
    transforms[wp.tid()] = wp.mat44d(
        wp.float64(1.0),
        wp.float64(0.0),
        wp.float64(0.0),
        wp.float64(0.0),
        wp.float64(0.0),
        wp.float64(1.0),
        wp.float64(0.0),
        wp.float64(0.0),
        wp.float64(0.0),
        wp.float64(0.0),
        wp.float64(1.0),
        wp.float64(0.0),
        x,
        wp.float64(0.0),
        wp.float64(0.0),
        wp.float64(1.0),
    )


def _load_base(stage):
    ovstage.population.open_usd(stage, TEST_BASE_PATH, ordinal=1)
    stage.advance_write_floor(1, ovstage.Scope.ALL).wait()


def _load_base_legacy(renderer):
    renderer.open_usd(TEST_BASE_PATH)
    renderer.reset()


def _read_xform(stage, paths, query, ordinal):
    attribute = paths.intern_token("omni:xform")
    with stage.read_attributes(query, [attribute], ovstage.OrdinalRange.latest(ordinal)) as read:
        group = read.fetch_next()
        values = np.from_dlpack(group.dlpack(0)).copy().reshape(1, 4, 4)
        stage.release_group(group)
    return values


def _read_xform_legacy(renderer, prim="/World/Plane"):
    tensor = renderer.read_attribute("omni:xform", [prim])
    return np.from_dlpack(tensor).reshape(1, 4, 4)


def _make_xform(x):
    matrix = np.eye(4, dtype=np.float64).reshape(1, 4, 4)
    matrix[0, 3, 0] = x
    return matrix


def test_bind_attribute_write_unbind(stage):
    """Create a scalar binding, write through it, and read back the result."""
    _load_base(stage)
    matrix = _make_xform(12.0)

    with ovstage.PathDictionary(stage) as paths:
        path_list = paths.create_path_list_from_strings(["/World/Plane"])

        # [snippet:doc-bind-attribute-write]
        query = stage.query_from_path_list(path_list)
        attribute = paths.intern_token("omni:xform")
        matrix_dtype = ovstage.numpy_to_dldatatype(matrix.dtype, lanes=16)
        matrix_tensor = ovstage.make_dltensor(matrix, dtype=matrix_dtype, shape=[1], ndim=1)
        stage.write_attribute(query, attribute, ordinal=2, tensors=matrix_tensor, is_array=False).wait()
        stage.advance_write_floor(2, ovstage.Scope.ALL).wait()
        # [/snippet:doc-bind-attribute-write]

        np.testing.assert_allclose(_read_xform(stage, paths, query, 2), matrix)
        query.release().wait()
        paths.destroy_path_list(path_list)


def test_bind_attribute_async_and_write_async(stage):
    """Create and write a binding asynchronously."""
    _load_base(stage)
    matrix = _make_xform(13.0)

    with ovstage.PathDictionary(stage) as paths:
        # [snippet:doc-bind-attribute-async]
        plane_filter = ovstage.Filter([ovstage.Predicate("usd-path", ovstage.FilterOp.IN, ["/World/Plane"])])
        query = stage.query(filter=plane_filter)
        query.wait()
        # [/snippet:doc-bind-attribute-async]

        # [snippet:doc-binding-write-async]
        attribute = paths.intern_token("omni:xform")
        matrix_dtype = ovstage.numpy_to_dldatatype(matrix.dtype, lanes=16)
        matrix_tensor = ovstage.make_dltensor(matrix, dtype=matrix_dtype, shape=[1], ndim=1)
        write_op = stage.write_attribute(query, attribute, ordinal=2, tensors=matrix_tensor, is_array=False)
        write_op.wait()
        stage.advance_write_floor(2, ovstage.Scope.ALL).wait()
        # [/snippet:doc-binding-write-async]

        np.testing.assert_allclose(_read_xform(stage, paths, query, 2), matrix)
        query.release().wait()


def test_bind_array_attribute(stage):
    """Bind a variable-length array attribute, write through it, and read it back."""
    _load_base(stage)
    points = np.array(
        [
            [-1.0, 0.0, -1.0],
            [1.0, 0.0, -1.0],
            [-1.0, 0.0, 1.0],
            [1.0, 0.0, 1.0],
        ],
        dtype=np.float32,
    )

    with ovstage.PathDictionary(stage) as paths:
        path_list = paths.create_path_list_from_strings(["/World/Plane"])
        query = stage.query_from_path_list(path_list)

        # [snippet:doc-bind-array-attribute]
        attribute = paths.intern_token("points")
        point_dtype = ovstage.numpy_to_dldatatype(points.dtype, lanes=3)
        point_tensor = ovstage.make_dltensor(points, dtype=point_dtype, shape=[4], ndim=1)
        stage.write_attribute(query, attribute, ordinal=2, tensors=point_tensor, is_array=True).wait()
        stage.advance_write_floor(2, ovstage.Scope.ALL).wait()
        # [/snippet:doc-bind-array-attribute]

        with stage.read_attributes(query, [attribute], ovstage.OrdinalRange.latest(2)) as read:
            group = read.fetch_next()
            values = np.from_dlpack(group.dlpack(0)).copy()
            stage.release_group(group)
        np.testing.assert_allclose(values, points)
        query.release().wait()
        paths.destroy_path_list(path_list)


def test_map_attribute_cpu(stage):
    """Map an attribute by name on CPU and verify the mutation."""
    _load_base(stage)

    with ovstage.PathDictionary(stage) as paths:
        path_list = paths.create_path_list_from_strings(["/World/Plane"])
        query = stage.query_from_path_list(path_list)
        attribute = paths.intern_token("omni:xform")

        # [snippet:doc-map-attribute-cpu]
        with stage.map_attribute(query, attribute, ordinal=2) as mapping:
            mapping.wait()
            group = mapping.fetch_next()
            matrices = np.from_dlpack(group.dlpack(0)).reshape(1, 4, 4)
            matrices[0, 3, 0] = 10.0
        stage.advance_write_floor(2, ovstage.Scope.ALL).wait()
        # [/snippet:doc-map-attribute-cpu]

        assert _read_xform(stage, paths, query, 2)[0, 3, 0] == 10.0
        query.release().wait()
        paths.destroy_path_list(path_list)


def test_map_bound_attribute_cpu(stage):
    """Map through an AttributeBinding rather than by name."""
    _load_base(stage)

    with ovstage.PathDictionary(stage) as paths:
        path_list = paths.create_path_list_from_strings(["/World/Plane"])

        # [snippet:doc-map-bound-attribute]
        query = stage.query_from_path_list(path_list)
        attribute = paths.intern_token("omni:xform")
        with stage.map_attribute(query, attribute, ordinal=2) as mapping:
            mapping.wait()
            group = mapping.fetch_next()
            matrices = np.from_dlpack(group.dlpack(0)).reshape(1, 4, 4)
            matrices[0, 3, 0] = 8.0
        stage.advance_write_floor(2, ovstage.Scope.ALL).wait()
        # [/snippet:doc-map-bound-attribute]

        assert _read_xform(stage, paths, query, 2)[0, 3, 0] == 8.0
        query.release().wait()
        paths.destroy_path_list(path_list)


def test_unmap_attribute_async(stage):
    """Unmap explicitly through the async API."""
    _load_base(stage)

    with ovstage.PathDictionary(stage) as paths:
        path_list = paths.create_path_list_from_strings(["/World/Plane"])
        query = stage.query_from_path_list(path_list)
        attribute = paths.intern_token("omni:xform")

        # [snippet:doc-unmap-attribute-async]
        mapping = stage.map_attribute(query, attribute, ordinal=2)
        mapping.wait()
        group = mapping.fetch_next()
        matrices = np.from_dlpack(group.dlpack(0)).reshape(1, 4, 4)
        matrices[0, 3, 0] = 9.0
        op = mapping.unmap()
        op.wait()
        stage.advance_write_floor(2, ovstage.Scope.ALL).wait()
        # [/snippet:doc-unmap-attribute-async]

        assert _read_xform(stage, paths, query, 2)[0, 3, 0] == 9.0
        query.release().wait()
        paths.destroy_path_list(path_list)


@pytest.mark.filterwarnings("ignore:.* is deprecated in ovrtx 0\\.4\\..*:DeprecationWarning")
def test_map_attribute_cuda(renderer):
    """Map an attribute on CUDA, edit it with Warp, and read back on CPU."""
    _load_base_legacy(renderer)

    # [snippet:doc-map-attribute-cuda]
    mapping = renderer.map_attribute(
        ["/World/Plane"],
        "omni:xform",
        dtype="float64",
        shape=(4, 4),
        device=ovrtx.Device.CUDA,
    )
    tensor = wp.from_dlpack(mapping.tensor, dtype=wp.mat44d)
    stream = wp.Stream(device=tensor.device)
    wp.launch(_set_xform_translation_x, dim=1, inputs=[tensor, wp.float64(6.0)], stream=stream)
    mapping.unmap(stream=stream.cuda_stream)
    # [/snippet:doc-map-attribute-cuda]

    assert _read_xform_legacy(renderer)[0, 3, 0] == 6.0


@pytest.mark.filterwarnings("ignore:.* is deprecated in ovrtx 0\\.4\\..*:DeprecationWarning")
def test_write_attribute_async_data_access_cuda(renderer):
    """Write a CUDA tensor with asynchronous data access and stream sync."""
    _load_base_legacy(renderer)
    cuda_tensor = wp.empty(1, dtype=wp.mat44d, device="cuda:0")
    stream = wp.Stream(device=cuda_tensor.device)
    wp.launch(_set_xform_translation_x, dim=1, inputs=[cuda_tensor, wp.float64(7.0)], stream=stream)

    # [snippet:doc-write-attribute-async-data-access]
    op = renderer.write_attribute_async(
        ["/World/Plane"],
        "omni:xform",
        cuda_tensor,
        data_access=ovrtx.DataAccess.ASYNC,
        cuda_stream=stream.cuda_stream,
    )
    assert op.wait() is True
    # [/snippet:doc-write-attribute-async-data-access]

    assert _read_xform_legacy(renderer)[0, 3, 0] == 7.0


def test_write_token_array_attribute(stage):
    """Write string data as a token array rather than a path relationship."""
    _load_base(stage)

    with ovstage.PathDictionary(stage) as paths:
        path_list = paths.create_path_list_from_strings(["/World/Plane"])
        query = stage.query_from_path_list(path_list)

        # [snippet:doc-write-token-array]
        attribute = paths.intern_token("omni:docTokens")
        token_ids = np.array([paths.intern_token("sensor"), paths.intern_token("validated")], dtype=np.uint64)
        stage.write_attribute(
            query,
            attribute,
            ordinal=2,
            tensors=token_ids,
            is_array=True,
            semantic=ovstage.AttributeSemantic.TOKEN_ID,
        ).wait()
        stage.advance_write_floor(2, ovstage.Scope.ALL).wait()
        # [/snippet:doc-write-token-array]

        with stage.read_attributes(query, [attribute], ovstage.OrdinalRange.latest(2)) as read:
            group = read.fetch_next()
            assert group.is_array
            assert group.tensor(0).dtype.lanes == 1
            assert group.attribute == attribute
            stage.release_group(group)
        query.release().wait()
        paths.destroy_path_list(path_list)

File: tests/docs/c/test_attribute_bindings.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: LicenseRef-NvidiaProprietary
//
// NVIDIA CORPORATION, its affiliates and licensors retain all intellectual
// property and proprietary rights in and to this material, related
// documentation and any modifications thereto. Any use, reproduction,
// disclosure or distribution of this material and related documentation
// without an express license agreement from NVIDIA CORPORATION or
// its affiliates is strictly prohibited.

// Tests for the ovstage "binding" pattern (query handle + interned attribute
// token) and zero-copy map/unmap writes. Mirrors the query-as-binding shape
// of tests/docs/python/test_attribute_bindings.py. The old ovrtx binding-
// handle model (ovrtx_create_attribute_binding / ovrtx_destroy_attribute_binding)
// is deprecated in 0.4; ovstage's query handles play that role directly.

#include <gtest/gtest.h>
#include "helpers.h"

#include <cstdint>
#include <cstring>
#include <string>
#include <vector>

namespace {

// Read /World/Plane's omni:xform translation X back for verification. Kept
// out of the snippet blocks — the docs render the write path only.
void read_xform_translation_x(ovstage_instance_t* stage,
                              char const* prim_path,
                              ovstage_ordinal_t ordinal,
                              double* out_x) {
    std::vector<uint8_t> bytes;
    docs_read_attribute(stage, prim_path, "omni:xform", ordinal, &bytes);
    ASSERT_GE(bytes.size(), sizeof(double) * 16u);
    // 4x4 matrix, USD row-vector convention — translation in the last row,
    // so translation.x is at flat index 3*4 + 0 = 12.
    double const* matrix = reinterpret_cast<double const*>(bytes.data());
    *out_x = matrix[12];
}

} // namespace

class AttributeBindingsTest : public DocsOvstageTestBase {
protected:
    DOCS_OVSTAGE_TEST_SUITE(AttributeBindingsTest)
};

TEST_F(AttributeBindingsTest, CreateWriteDestroyBinding) {
    docs_load_base();
    if (HasFatalFailure()) return;

    // [snippet:doc-create-attribute-binding-c]
    // The ovstage "binding" is the pair (query handle, interned attribute
    // token). The query identifies the target prims; the token identifies the
    // attribute. Both are reserved synchronously and reusable across many
    // writes/reads until released.
    path_dictionary_instance_t* pd = ovstage_get_path_dictionary(stage_);
    ovx_string_t prim_path = ovx_str("/World/Plane");
    ovx_primpath_list_t path_list{};
    ASSERT_EQ(path_dictionary_create_path_list_from_strings(pd, &prim_path, 1, &path_list).status,
              OVX_API_SUCCESS);

    ovstage_query_handle_t query_handle = OVSTAGE_INVALID_QUERY_HANDLE;
    ASSERT_EQ(ovstage_query_from_path_list(stage_, path_list, &query_handle), OVSTAGE_OK)
        << format_ovstage_last_error();

    ovx_string_t attr_str = ovx_str("omni:xform");
    ovx_token_t attr_token{};
    ASSERT_EQ(path_dictionary_create_tokens_from_strings(pd, &attr_str, 1, &attr_token).status,
              OVX_API_SUCCESS);
    // [/snippet:doc-create-attribute-binding-c]

    // [snippet:doc-write-bound-attribute-c]
    // Write through the (query, token) binding. omni:xform is a per-prim 4x4
    // double matrix — shape=[1], lanes=16, with OVSTAGE_SEMANTIC_MATRIX.
    // Translation lives in the last row (USD row-vector convention).
    double matrix[16] = {
        1.0,  0.0,  0.0, 0.0,
        0.0,  1.0,  0.0, 0.0,
        0.0,  0.0,  1.0, 0.0,
        14.0, 0.0,  0.0, 1.0,
    };
    int64_t write_shape[1] = {1};
    DLTensor write_tensor{};
    write_tensor.data = matrix;
    write_tensor.device = {kDLCPU, 0};
    write_tensor.ndim = 1;
    write_tensor.dtype = {kDLFloat, 64, 16};
    write_tensor.shape = write_shape;

    ovstage_write_data_t write_data{};
    write_data.tensors = &write_tensor;
    write_data.tensor_count = 1;
    write_data.is_array = false;
    write_data.semantic = OVSTAGE_SEMANTIC_MATRIX;

    ovx_string_or_token_t attr_ref{};
    attr_ref.token = attr_token;

    ovstage_enqueue_result_t wq = ovstage_write_attribute(
        stage_, query_handle, attr_ref, /*ordinal=*/2, write_data, OVSTAGE_PRIM_MODE_UPSERT);
    ASSERT_EQ(wq.status, OVSTAGE_OK) << format_ovstage_last_error();
    docs_wait_ovstage_no_errors(stage_, wq.op_index);
    docs_ovstage_advance_write_floor(stage_, 2);
    // [/snippet:doc-write-bound-attribute-c]

    double x = 0.0;
    read_xform_translation_x(stage_, "/World/Plane", 2, &x);
    EXPECT_DOUBLE_EQ(x, 14.0);

    // [snippet:doc-destroy-attribute-binding-c]
    // Release the query handle (per-handle-ordered: waits for in-flight
    // reads/writes to complete) and drop the path-list refcount.
    ovstage_release_query(stage_, query_handle);
    path_dictionary_release_path_list_reference(pd, path_list);
    // [/snippet:doc-destroy-attribute-binding-c]
}

TEST_F(AttributeBindingsTest, MapAndUnmapAttribute) {
    docs_load_base();
    if (HasFatalFailure()) return;

    // Set up the (query, token) binding — same shape as CreateWriteDestroyBinding.
    path_dictionary_instance_t* pd = ovstage_get_path_dictionary(stage_);
    ovx_string_t prim_path = ovx_str("/World/Plane");
    ovx_primpath_list_t path_list{};
    ASSERT_EQ(path_dictionary_create_path_list_from_strings(pd, &prim_path, 1, &path_list).status,
              OVX_API_SUCCESS);
    ovstage_query_handle_t query_handle = OVSTAGE_INVALID_QUERY_HANDLE;
    ASSERT_EQ(ovstage_query_from_path_list(stage_, path_list, &query_handle), OVSTAGE_OK)
        << format_ovstage_last_error();
    ovx_string_t attr_str = ovx_str("omni:xform");
    ovx_token_t attr_token{};
    ASSERT_EQ(path_dictionary_create_tokens_from_strings(pd, &attr_str, 1, &attr_token).status,
              OVX_API_SUCCESS);

    // [snippet:doc-map-attribute-cpu-c]
    // Zero-copy write via map/unmap: map_attribute hands back writable
    // storage for each matched prim; the caller fills it, then unmap_attribute
    // commits and releases the map session.
    ovx_string_or_token_t attr_ref{};
    attr_ref.token = attr_token;
    ovstage_map_desc_t map_desc{};
    map_desc.attribute = attr_ref;
    map_desc.dtype = {kDLFloat, 64, 16};
    map_desc.semantic = OVSTAGE_SEMANTIC_MATRIX;
    map_desc.prim_mode = OVSTAGE_PRIM_MODE_UPSERT;

    ovstage_map_handle_t map_handle = OVSTAGE_INVALID_MAP_HANDLE;
    ovstage_enqueue_result_t mq = ovstage_map_attribute(
        stage_, query_handle, &map_desc, /*ordinal=*/2,
        /*element_sizes=*/nullptr, /*element_count=*/0, &map_handle);
    ASSERT_EQ(mq.status, OVSTAGE_OK) << format_ovstage_last_error();
    docs_wait_ovstage_no_errors(stage_, mq.op_index);

    ovstage_map_group_t map_group{};
    ASSERT_EQ(ovstage_fetch_map_next(stage_, map_handle, OVSTAGE_TIMEOUT_INFINITE, &map_group),
              OVSTAGE_OK);
    double* matrix = static_cast<double*>(map_group.data.tensors[0].data);
    matrix[0] = 1.0;
    matrix[5] = 1.0;
    matrix[10] = 1.0;
    matrix[15] = 1.0;
    matrix[12] = 15.0; // translation.x under USD row-vector convention

    // Commit all pending groups and release the map handle.
    ovstage_cuda_sync_t no_sync{};
    ovstage_enqueue_result_t uq = ovstage_unmap_attribute(stage_, map_handle, no_sync);
    ASSERT_EQ(uq.status, OVSTAGE_OK) << format_ovstage_last_error();
    docs_wait_ovstage_no_errors(stage_, uq.op_index);
    docs_ovstage_advance_write_floor(stage_, 2);
    // [/snippet:doc-map-attribute-cpu-c]

    double x = 0.0;
    read_xform_translation_x(stage_, "/World/Plane", 2, &x);
    EXPECT_DOUBLE_EQ(x, 15.0);

    ovstage_release_query(stage_, query_handle);
    path_dictionary_release_path_list_reference(pd, path_list);
}
